pkg/login/sessions.go

```diff
@@ -0,0 +1,294 @@
+		return err
+	}
+	if err := json.NewDecoder(bytes.NewReader(obj.Bytes)).Decode(&dest); err != nil {
+		return fmt.Errorf("json.Decode() failed: %v", err)
+	}
+	return nil
+}
+
+func (x *Store) writeMap(ctx context.Context, src *map[SessionID]Identity) error {
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(src); err != nil {
+		return fmt.Errorf("json.Decode() failed: %v", err)
+	}
+	if err := x.cache.PutObject(ctx, DataFile, storage.NewJSONObject(buf.Bytes())); err != nil {
+		return fmt.Errorf("x.cache.PutObject() failed: %v", err)
+	}
+	return nil
+}
+
+var ErrInvalidSession = errors.New("invalid session")
+var ErrInvalidSignature = errors.New("invalid signature")
+
+const SessionFormat = "V-i-r-t-u-a-l--V-G-O--%016x%016x%016x%016x%016x%016x"
+
+// Sessions can be encoded as url safe strings and embedded into cookies or http headers.
+// Session id's are used look up user access roles.
+// Each session should get a unique id.
+type Session struct {
+	// Id is a unique random session id
+	ID SessionID
+
+	// ExpiresNanos is the time in second since epoch that this session expires
+	ExpiresNanos uint64
+}
+
+type SessionID uint64
+
+// Encodes the session into a url safe string.
+// The secret is used to cryptographically sign the session data.
+func (x *Session) Encode(secret Secret) string {
+	hash := x.makeSignature(secret)
+	return fmt.Sprintf(SessionFormat, hash[0], hash[1], hash[2], hash[3], x.ID, x.ExpiresNanos)
+}
+
+// DecodeCookie reads session data stored in the cookie.
+// Secret is used to validate the cookie's signature.
+func (x *Session) DecodeCookie(secret Secret, src *http.Cookie) error {
+	return x.Decode(secret, src.Value)
+}
+
+// Decode reads session from a string.
+// Secret is used to validate the session signature.
+// * ErrInvalidSession is returned when the session code not be decoded.
+// * ErrInvalidSignature is returned when the session was read successfully, but signature was invalid.
+// * ErrSessionExpired is returned when the session was read and has a valid signature, but the session is expired.
+// If ErrInvalidSignature or ErrSessionExpired, the session data is still written to this object.
+func (x *Session) Decode(secret Secret, value string) error {
+	// read the cookie
+	var sig [4]uint64
+	_, err := fmt.Sscanf(value, SessionFormat,
+		&sig[0], &sig[1], &sig[2], &sig[3], &x.ID, &x.ExpiresNanos)
+	if err != nil {
+		return ErrInvalidSession
+	}
+
+	switch {
+	case x.makeSignature(secret) != sig:
+		return ErrInvalidSignature
+	case uint64(time.Now().UnixNano()) >= x.ExpiresNanos:
+		return ErrSessionExpired
+	default:
+		return nil
+	}
+}
+
+func (x *Session) makeSignature(secret Secret) [4]uint64 {
+	str := fmt.Sprintf("%v|%v|%v", secret, x.ID, x.ExpiresNanos)
+	sum := sha256.Sum256([]byte(str))
+	sumReader := bytes.NewReader(sum[:])
+	var hash [4]uint64
+	for i := range hash {
+		binary.Read(sumReader, binary.LittleEndian, &hash[i])
+	}
```

Comment:
I think it might be more idiomatic to use `BigEndian.Uint64` from [`encoding/binary`](https://golang.org/pkg/encoding/binary/#ByteOrder).

Reply:
I actually don't really know the difference here. What makes it more idomatic?

Reply:
Something like this, modified for Uint64 instead of 16:
```
b := []byte{0x12, 0x34, 0x56, 0x78}
x1 := binary.BigEndian.Uint16(b[0:])
x2 := binary.BigEndian.Uint16(b[2:])
```

Saves the overhead of instantiating a Reader.

Also, BigEndian, for overall endian-ness consistency. Little Endian turns `0x12345678` into `[0x21, 0x43, 0x65, 0x87]`, if I'm not mistaken.

Reply:
:ok: sounds good! That looks much cleaner! :bow: